bin/wayland-launch

```diff
@@ -2,8 +2,27 @@
 
 set -x
 
-real_wayland=$(dirname "$XDG_RUNTIME_DIR")/${WAYLAND_DISPLAY:-wayland-0}
-while [ ! -O "${real_wayland}" ]; do echo waiting for Wayland socket; sleep 4; done
+if ! snapctl is-connected wayland
+then
+  echo "Wayland interface not connected!"
+  exit 1
+fi
+
+real_xdg_runtime_dir=$(dirname "${XDG_RUNTIME_DIR}")
+real_wayland=${real_xdg_runtime_dir}/${WAYLAND_DISPLAY:-wayland-0}
+
+while [ ! -O "${real_wayland}" ]
+do
+  # On core systems mir-kiosk may also need to create the host XDG_RUNTIME_DIR
+  while [ ! -O "${real_xdg_runtime_dir}" ]
+  do
+    echo waiting for host XDG_RUNTIME_DIR
+    inotifywait --event create $(dirname "${real_xdg_runtime_dir}") || sleep 4
+  done
+
+  echo waiting for Wayland socket
+  inotifywait --event create $(dirname "${real_wayland}") || sleep 4
+done
```

Comment:
IIUC this will wait indefinitely unless `XDG_RUNTIME_DIR` gets created and then something gets created inside? `inotifywait` has no default timeout, and even if it did, the `while` around it will just loop forever. And the `echo`s will print in a loop. All that by design?

Also, `until` may be used in place of `while !`.

Reply:
FWIW this is what I used elsewhere:
```bash
timeout 5 sh -c 'until [ -S $XDG_RUNTIME_DIR/wayland-0 ]; do sleep 1; done' || (
  echo "ERROR mir-kiosk failed to create its socket."
  exit 1
)
```

Reply:
Where are you doing that?

Reply:
> IIUC this will wait indefinitely unless `XDG_RUNTIME_DIR` gets created and then something gets created inside?

Yes. That's essentially the existing behavior.

Using timeout has the same issue that using inotifywait has: existing snaps may not stage it.

I can come up with something more robust. But a stage snap would be a better approach. This is just a clean-up of the existing code until I get around to a rework.

Reply:
> Using timeout has the same issue that using inotifywait has: existing snaps may not stage it.

All the core* bases ship it:
```bash
$ ls /snap/core*/current/usr/bin/timeout
/snap/core18/current/usr/bin/timeout  /snap/core20/current/usr/bin/timeout  /snap/core/current/usr/bin/timeout
```

> Yes. That's essentially the existing behavior.

Sure, but do you think it's the right behaviour? Should we not bail out after a timeout, and let the consumer deal with that (by having systemd restart it, or realize that something's gone wrong)?

Reply:
> All the core* bases ship it:

Shipping is not enough. In a `base: core20` snap (mir-test-tools as per PR):
```plain
$ timeout --help
bash: /usr/bin/timeout: Permission denied
```

In a `base: core18` snap (mir-kiosk from edge):
```plain
$ timeout --help
bash: /usr/bin/timeout: Permission denied
```

Reply:
In any case, are you blocking this PR because you don't like *existing* behavior it doesn't change?

Reply:
> bash: /usr/bin/timeout: Permission denied

Bleh. Sounds like a bug in snapd.

> In any case, are you blocking this PR because you don't like existing behavior it doesn't change?

Not blocking, discussing :)